src/Features/CSharp/Portable/ConvertForEachToFor/CSharpConvertForEachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,161 @@
+        {
+            var foreachStatement = token.Parent.FirstAncestorOrSelf<ForEachStatementSyntax>();
+            if (foreachStatement == null)
+            {
+                return null;
+            }
+
+            // support refactoring only if caret is in between "foreach" and ")"
+            var scope = TextSpan.FromBounds(foreachStatement.ForEachKeyword.Span.Start, foreachStatement.CloseParenToken.Span.End);
+            if (!scope.IntersectsWith(selection))
+            {
+                return null;
+            }
+
+            // check whether there is any comments between foreach and ) tokens
+            // if they do, we don't support conversion.
+            foreach (var trivia in foreachStatement.DescendantTrivia(n => n == foreachStatement || scope.Contains(n.FullSpan)))
+            {
+                if (trivia.Span.End <= scope.Start ||
+                    scope.End <= trivia.Span.Start)
+                {
+                    continue;
+                }
+
+                if (trivia.Kind() != SyntaxKind.WhitespaceTrivia &&
+                    trivia.Kind() != SyntaxKind.EndOfLineTrivia)
+                {
+                    // we don't know what to do with these comments
+                    return null;
+                }
+            }
+
+            return foreachStatement;
+        }
+
+        protected override bool ValidLocation(ForEachInfo foreachInfo)
+        {
+            if (!foreachInfo.RequireCollectionStatement)
+            {
+                return true;
+            }
+
+            // for now, we don't support converting in embeded statement if 
+            // new local declaration for collection is required.
+            // we can support this by using Introduce local variable service
+            // but the service is not currently written in a way that can be
+            // easily reused here.
+            return foreachInfo.ForEachStatement.Parent.IsKind(SyntaxKind.Block);
+        }
+
+        protected override (SyntaxNode start, SyntaxNode end) GetForEachBody(ForEachStatementSyntax foreachStatement)
+            => (foreachStatement.Statement, foreachStatement.Statement);
+
+        protected override void ConvertToForStatement(
+            SemanticModel model, ForEachInfo foreachInfo, SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var generator = editor.Generator;
+            var foreachStatement = foreachInfo.ForEachStatement;
+
+            var foreachCollectionExpression = foreachStatement.Expression;
+            var collectionVariable = GetCollectionVariableName(
+                model, generator, foreachInfo, foreachCollectionExpression, cancellationToken);
+
+            var collectionStatementType = foreachInfo.RequireExplicitCastInterface
+                ? generator.GetTypeExpression(foreachInfo.Options, foreachInfo.ExplicitCastInterface) :
```

Comment:
IDE almost always aligns ? and :  when wrapping ternaries.  can you please keep this style like that?